### PR TITLE
Fixed issue with spreading a generic call expression into generic JSX and gather intra expression inference sites from spread expressions

### PR DESCRIPTION
**Pull request by Andarist**
This PR:
- reverts a [reverted PR](https://github.com/microsoft/TypeScript/pull/51580) to fix https://github.com/microsoft/TypeScript/issues/51577 again
- gathers intra-expression inferences sites from spread expressions as lack of this was the reason why my original PR was reverted. Note though that regular call expressions were never adding those from spread expressions but, somewhat accidentally, JSX was gathering those. This PR gathers them now in both cases.

cc @jakebailey @weswigham

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -30130,7 +30130,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     hasComputedNumberProperty = false;
                     hasComputedSymbolProperty = false;
                 }
-                const type = getReducedType(checkExpression(memberDecl.expression));
+                const type = getReducedType(checkExpression(memberDecl.expression, checkMode & CheckMode.Inferential));
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply src/compiler/checker.ts:30133]: Dub question, but, why not just pass check mode alone? (I wish we made the parameter required everywhere...)

[the remark this answers is not in the document]
Comment (Andarist) [reply src/compiler/checker.ts:30133]: That’s a question i can’t answer. I just figured out that the original issue was caused by the fact that it **was** passed down and that it was inconsistent with regular calls. So I unified both paths to work the same.

that other issue with spreads and intra expression inferences was *accidentally* discovered - it wasn’t exactly what was reported as broken for lack of this feature in JSX. It just was an example of a workaround and when I was fixing that issue I just copy pasted all of the examples from the issue. If not that, it’s likely that we would never know about this being broken - although in context of JSX there are some people who prefer to spread object literals (which is not common with regular calls, at least not in such a simple form). @weswigham mentioned in the comment that this case looked alright to him and that we don't want to break it… so here we are :)

**Review by weswigham — APPROVED** [review]
Filtering the `checkMode` we pass forward is a bit odd, but it wouldn't be the first place we, at least implicitly, do so (usually we implicitly replace the active checkMode with `Normal`). This is probably going to cause some edge cases to change in object literal contextually typed function member inference (since the `Contextual` checkmode isn't being forwarded anymore), but maybe not. Let's just run the tests, since I can't come up with an example that'd break offhand.

**Comment by sandersn** [comment]
@weswigham is this good for 5.1?

**Comment by Andarist** [comment]
This PR mainly unifies behavior of JSX and regular calls so I think that it should go in sooner than later.

**Review by jakebailey — APPROVED** [review]
Seems like this is ready, then? Surprised it wasn't merged unless I've missed something.
